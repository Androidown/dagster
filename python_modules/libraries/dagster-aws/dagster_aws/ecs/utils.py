import hashlib
import re
from typing import Any, Mapping

from dagster._core.remote_representation.origin import RemoteJobOrigin

from .tasks import DagsterEcsTaskDefinitionConfig


def sanitize_family(family):
    # Trim the location name and remove special characters
    return re.sub(r"[^\w^\-]", "", family)[:255]


def _get_family_hash(name):
    m = hashlib.sha1()
    m.update(name.encode("utf-8"))
    name_hash = m.hexdigest()[:8]
    return f"{name[:55]}_{name_hash}"


def get_task_definition_family(
    prefix: str,
    job_origin: RemoteJobOrigin,
) -> str:
    job_name = job_origin.job_name
    repo_name = job_origin.repository_origin.repository_name
    location_name = job_origin.repository_origin.code_location_origin.location_name

    assert len(prefix) < 32

    # Truncate the location name if it's too long (but add a unique suffix at the end so that no matter what it's unique)
    # Relies on the fact that org name and deployment name are always <= 64 characters long to
    # stay well underneath the 255 character limit imposed by ECS

    final_family = f"{prefix}_{_get_family_hash(location_name)}_{_get_family_hash(repo_name)}_{_get_family_hash(job_name)}"

    assert len(final_family) <= 255

    return sanitize_family(final_family)


def task_definitions_match(
    desired_task_definition_config: DagsterEcsTaskDefinitionConfig,
    existing_task_definition: Mapping[str, Any],
    container_name: str,
) -> bool:
    if not any(
        [
            container["name"] == container_name
            for container in existing_task_definition["containerDefinitions"]
        ]
    ):
        return False

    existing_task_definition_config = DagsterEcsTaskDefinitionConfig.from_task_definition_dict(
        existing_task_definition, container_name
    )

    # sidecars are checked separately below
    match_without_sidecars = existing_task_definition_config._replace(
        sidecars=[],
    ) == desired_task_definition_config._replace(
        sidecars=[],
    )
    if not match_without_sidecars:
        return False

    # Just match sidecars on certain fields
    if not [
        (
            sidecar["name"],
            sidecar["image"],
            sidecar.get("environment", []),
            sidecar.get("secrets", []),
        )
        for sidecar in existing_task_definition_config.sidecars
    ] == [
        (
            sidecar["name"],
            sidecar["image"],
            sidecar.get("environment", []),
            sidecar.get("secrets", []),
        )
        for sidecar in desired_task_definition_config.sidecars
    ]:
        return False

    return True


def get_task_logs(ecs, logs_client, cluster, task_arn, container_name, limit=10):
    task = ecs.describe_tasks(cluster=cluster, tasks=[task_arn]).get("tasks")[0]

    task_definition_arn = task.get("taskDefinitionArn")
    task_definition = ecs.describe_task_definition(taskDefinition=task_definition_arn).get(
        "taskDefinition"
    )

    matching_container_definitions = [
        container_definition
        for container_definition in task_definition.get("containerDefinitions", [])
        if container_definition["name"] == container_name
    ]
    if not matching_container_definitions:
        raise Exception(f"Could not find container with name {container_name}")

    container_definition = matching_container_definitions[0]

    log_options = container_definition.get("logConfiguration", {}).get("options", {})
    log_group = log_options.get("awslogs-group")
    log_stream_prefix = log_options.get("awslogs-stream-prefix")

    if not log_group or not log_stream_prefix:
        return []

    container_name = container_definition.get("name")
    task_id = task_arn.split("/")[-1]

    log_stream = f"{log_stream_prefix}/{container_name}/{task_id}"

    events = logs_client.get_log_events(
        logGroupName=log_group, logStreamName=log_stream, limit=limit
    ).get("events")

    return [event.get("message") for event in events]
