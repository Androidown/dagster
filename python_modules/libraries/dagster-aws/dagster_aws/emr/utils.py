import copy

import dagster._check as check


def subset_run_config(run_config, op_name):
    """Drops op config for ops other than solid_name; this subsetting is required when
    executing a single op on EMR to pass config validation.
    """
    check.dict_param(run_config, "run_config")
    check.str_param(op_name, "solid_name")

    subset = copy.deepcopy(run_config)
    if "ops" in subset:
        solid_config_keys = list(subset["ops"].keys())
        for key in solid_config_keys:
            if key != op_name:
                del subset["ops"][key]
    return subset
