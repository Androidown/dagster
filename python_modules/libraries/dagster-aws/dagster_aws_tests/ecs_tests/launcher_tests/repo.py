import dagster


@dagster.op
def node(_):
    pass


@dagster.job
def job():
    node()


@dagster.repository
def repository():
    return {"jobs": {"job": job}}
