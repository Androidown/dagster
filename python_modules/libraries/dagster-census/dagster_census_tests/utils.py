def get_sync_data():
    return {
        "status": "success",
        "data": {
            "id": 61,
            "label": None,
            "schedule_frequency": "never",
            "schedule_day": None,
            "schedule_hour": None,
            "schedule_minute": None,
            "created_at": "2021-10-22T00:40:11.246Z",
            "updated_at": "2021-10-22T00:43:44.173Z",
            "operation": "upsert",
            "paused": False,
            "status": "Ready",
            "lead_union_insert_to": None,
            "trigger_on_dbt_cloud_rebuild": False,
            "field_behavior": "specific_properties",
            "field_normalization": None,
            "source_attributes": {
                "connection_id": 15,
                "object": {
                    "type": "model",
                    "id": 15,
                    "name": "braze_test",
                    "created_at": "2021-10-11T20:52:58.293Z",
                    "updated_at": "2021-10-14T23:15:18.508Z",
                    "query": "select email, cast('random' as VARCHAR(2000)) as random_prop",
                },
            },
            "destination_attributes": {"connection_id": 15, "object": "user"},
            "mappings": [
                {
                    "from": {"type": "column", "data": "EMAIL"},
                    "to": "external_id",
                    "is_primary_identifier": True,
                    "generate_field": False,
                    "preserve_values": False,
                    "operation": None,
                },
                {
                    "from": {
                        "type": "constant_value",
                        "data": {"value": "test", "basic_type": "text"},
                    },
                    "to": "first_name",
                    "is_primary_identifier": False,
                    "generate_field": False,
                    "preserve_values": False,
                    "operation": None,
                },
            ],
        },
    }


def get_source_data():
    return {
        "status": "success",
        "data": {
            "id": 15,
            "name": "Snowflake - xxxxxxx.us-east-1",
            "label": None,
            "type": "snowflake",
            "last_test_succeeded": None,
            "last_tested_at": None,
            "connection_details": {
                "account": "xxxxxxx.us-east-1",
                "user": "DEV",
                "warehouse": "TEST",
                "use_keypair": False,
            },
            "read_only_connection": False,
        },
    }


def get_destination_data():
    return {
        "status": "success",
        "data": {
            "id": 15,
            "name": "Braze",
            "connection_details": {"instance_url": "https://rest.iad-03.braze.com"},
            "objects": [
                {
                    "label": "User",
                    "full_name": "user",
                    "allow_custom_fields": True,
                    "allow_case_sensitive_field_names": True,
                },
                {
                    "label": "Event",
                    "full_name": "event",
                    "allow_custom_fields": True,
                    "allow_case_sensitive_field_names": True,
                },
            ],
        },
    }


def get_sync_run_data():
    return {
        "status": "success",
        "data": {
            "id": 94,
            "sync_id": 52,
            "source_record_count": 1,
            "records_processed": 1,
            "records_updated": 1,
            "records_failed": 0,
            "records_invalid": 0,
            "created_at": "2021-10-20T02:51:07.546Z",
            "updated_at": "2021-10-20T02:52:29.236Z",
            "completed_at": "2021-10-20T02:52:29.234Z",
            "scheduled_execution_time": None,
            "error_code": None,
            "error_message": None,
            "error_detail": None,
            "status": "completed",
            "canceled": False,
            "full_sync": True,
            "sync_trigger_reason": {
                "ui_tag": "Manual",
                "ui_detail": "Manually triggered by test@getcensus.com",
            },
        },
    }


def get_sync_trigger_data():
    return {"status": "success", "data": {"sync_run_id": 94}}
