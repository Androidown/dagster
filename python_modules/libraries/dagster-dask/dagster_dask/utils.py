# pyright: reportPrivateImportUsage=none

# NOTE (2022-12-12): We ignore private attribute access errors because dask does not correctly mark
# the symbols we are accessing as public (though they are intending to). Should be fixed in later
# releases of dask. See: https://github.com/dask/dask/issues/9710

import re

import dask.dataframe as dd
from dagster import Any, Bool, Field, Float, Int, Permissive, Shape, String


def normalize_column_names(df: dd.DataFrame, enabled) -> dd.DataFrame:
    if enabled:
        df.columns = normalize_names(df.columns)

    return df


def normalize_names(names):
    # Based on https://stackoverflow.com/a/1176023
    camel_to_snake1 = re.compile("(.)([A-Z][a-z]+)")
    camel_to_snake2 = re.compile("([a-z0-9])([A-Z])")

    def normalize(name):
        if not name:
            return name
        name = camel_to_snake1.sub(r"\1_\2", name)
        return camel_to_snake2.sub(r"\1_\2", name).lower()

    return [normalize(n) for n in names]


DataFrameUtilities = {
    "drop": {
        "function": dd.DataFrame.drop,
        "options": Field(
            Shape(
                {
                    "labels": Field(
                        Any,
                        is_required=False,
                        description="Index or column labels to drop.",
                    ),
                    "axis": Field(
                        Any,
                        is_required=False,
                        description="Drop labesl from index or columns.",
                    ),
                    "columns": Field(
                        Any,
                        is_required=False,
                        description="Equivalent to labels with axis=1.",
                    ),
                    "errors": Field(
                        String,
                        is_required=False,
                        description='If "ignore", supress errors.',
                    ),
                }
            ),
            is_required=False,
            description="Drop specified labels from rows or columns.",
        ),
    },
    "sample": {
        "function": dd.DataFrame.sample,
        "options": Field(
            Shape(
                {
                    "frac": Field(
                        Float,
                        is_required=False,
                        description="Fraction of axis items to return.",
                    ),
                    "replace": Field(
                        Bool,
                        is_required=False,
                        description="Sample with or without replacement.",
                    ),
                    "random_state": Field(
                        Int,
                        is_required=False,
                        description="Create RandomState with this as the seed.",
                    ),
                }
            ),
            is_required=False,
            description="Random sample of items.",
        ),
    },
    "reset_index": {
        "function": dd.DataFrame.reset_index,
        "options": Field(
            Shape(
                {
                    "drop": Field(
                        Bool,
                        is_required=False,
                        description="Do not try to insert index into dataframe columns.",
                    ),
                }
            ),
            is_required=False,
            description="Reset the index to the default index.",
        ),
    },
    "set_index": {
        "function": dd.DataFrame.set_index,
        "options": Field(
            Permissive(
                {
                    "other": Field(
                        String,
                        is_required=True,
                        description="Set index to specified column.",
                    ),
                    "drop": Field(
                        Bool,
                        is_required=False,
                        description="Delete columns to be used as the new index.",
                    ),
                    "sorted": Field(
                        Bool,
                        is_required=False,
                        description="If the index column is already sorted in increasing order.",
                    ),
                    "divisions": Field(
                        Any,
                        is_required=False,
                        description=(
                            "Known values on which to separate index values of the partitions."
                        ),
                    ),
                }
            ),
            is_required=False,
            description="Set the DataFrame index using an existing column.",
        ),
    },
    "repartition": {
        "function": dd.DataFrame.repartition,
        "options": Field(
            Shape(
                {
                    "divisions": Field(
                        Any,
                        is_required=False,
                        description="List of partitions to be used.",
                    ),
                    "npartitions": Field(
                        Int,
                        is_required=False,
                        description="Number of partitions of output.",
                    ),
                    "partition_size": Field(
                        Any,
                        is_required=False,
                        description="Max number of bytes of memory for each partition.",
                    ),
                    "freq": Field(
                        String,
                        is_required=False,
                        description="A period on which to partition timeseries data.",
                    ),
                    "force": Field(
                        Bool,
                        is_required=False,
                        description="Allows the expansion of the existing divisions.",
                    ),
                }
            ),
            is_required=False,
            description="Repartition dataframe along new divisions.",
        ),
    },
    "normalize_column_names": {
        "function": normalize_column_names,
        "options": Field(
            Bool,
            is_required=False,
            description="Lowercase and convert CamelCase to snake_case on column names.",
        ),
    },
}


def apply_utilities_to_df(df, config):
    for util_name, util_spec in DataFrameUtilities.items():
        if util_name in config:
            util_func = util_spec["function"]
            util_opts = config[util_name]

            if isinstance(util_opts, dict):
                df = util_func(df, **util_opts)
            elif isinstance(util_opts, list):
                df = util_func(df, *util_opts)
            else:
                df = util_func(df, util_opts)

    return df
