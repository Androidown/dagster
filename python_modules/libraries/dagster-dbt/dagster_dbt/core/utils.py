from collections import deque
from concurrent.futures import Future, ThreadPoolExecutor, TimeoutError
from typing import Any, Callable, Iterator, List, TypeVar, Union


def get_future_completion_state_or_err(futures: List[Union[Future, Any]]) -> bool:
    """Given a list of futures (and potentially other objects), return True if all futures are completed.
    If any future has an exception, raise the exception.
    """
    for future in futures:
        if not isinstance(future, Future):
            continue

        if not future.done():
            return False

        exception = future.exception()
        if exception:
            raise exception
    return True


T = TypeVar("T")
P = TypeVar("P")


def imap(
    executor: ThreadPoolExecutor,
    iterable: Iterator[T],
    func: Callable[[T], P],
) -> Iterator[P]:
    """A version of `concurrent.futures.ThreadpoolExecutor.map` which tails the input iterator in
    a separate thread. This means that the map function can begin processing and yielding results from
    the first elements of the iterator before the iterator is fully consumed.

    Args:
        executor: The ThreadPoolExecutor to use for parallel execution.
        iterable: The iterator to apply the function to.
        func: The function to apply to each element of the iterator.
    """
    work_queue: deque[Future] = deque([])

    # create a small task which waits on the iterator
    # and enqueues work items as they become available
    def _apply_func_to_iterator_results(iterable: Iterator[T]) -> None:
        for arg in iterable:
            work_queue.append(executor.submit(func, arg))

    enqueuing_task = executor.submit(
        _apply_func_to_iterator_results,
        iterable,
    )

    while True:
        if (not enqueuing_task or enqueuing_task.done()) and len(work_queue) == 0:
            break

        if len(work_queue) > 0:
            current_work_item = work_queue[0]

            try:
                yield current_work_item.result(timeout=0.1)
                work_queue.popleft()
            except TimeoutError:
                pass

    # Ensure any exceptions from the enqueuing task processing the iterator are raised,
    # after all work items have been processed.
    exc = enqueuing_task.exception()
    if exc:
        raise exc


def exhaust_iterator_and_yield_results_with_exception(iterable: Iterator[T]) -> Iterator[T]:
    """Fully exhausts an iterator and then yield its results. If the iterator raises an exception,
    raise that exception at the position in the iterator where it was originally raised.

    This is useful if we want to exhaust an iterator, but don't want to discard the elements
    that were yielded before the exception was raised.
    """
    results = []
    caught_exception = None
    try:
        for result in iterable:
            results.append(result)
    except Exception as e:
        caught_exception = e

    yield from results
    if caught_exception:
        raise caught_exception
