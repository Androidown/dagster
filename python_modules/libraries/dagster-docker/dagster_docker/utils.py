from dagster import (
    Field,
    StringSource,
    _check as check,
)
from dagster._utils.merger import merge_dicts
from docker_image import reference

from .container_context import DOCKER_CONTAINER_CONTEXT_SCHEMA

DOCKER_CONFIG_SCHEMA = merge_dicts(
    {
        "image": Field(
            StringSource,
            is_required=False,
            description="The docker image to be used if the repository does not specify one.",
        ),
        "network": Field(
            StringSource,
            is_required=False,
            description=(
                "Name of the network to which to connect the launched container at creation time"
            ),
        ),
    },
    DOCKER_CONTAINER_CONTEXT_SCHEMA,
)


def validate_docker_config(network, networks, container_kwargs):
    if network:
        check.invariant(not networks, "cannot set both `network` and `networks`")

    if container_kwargs:
        if "image" in container_kwargs:
            raise Exception(
                "'image' cannot be used in 'container_kwargs'. Use the 'image' config key instead."
            )

        if "environment" in container_kwargs:
            raise Exception(
                "'environment' cannot be used in 'container_kwargs'. Use the 'env_vars' config key"
                " instead."
            )

        if "network" in container_kwargs:
            raise Exception(
                "'network' cannot be used in 'container_kwargs'. Use the 'networks' config key"
                " instead."
            )


def validate_docker_image(docker_image):
    try:
        # validate that the docker image name is valid
        reference.Reference.parse(docker_image)
    except Exception as e:
        raise Exception(f"Docker image name {docker_image} is not correctly formatted") from e
