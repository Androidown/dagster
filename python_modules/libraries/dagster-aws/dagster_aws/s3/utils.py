from typing import Optional, Union

import boto3
import dagster._check as check
from botocore.handlers import disable_signing

from ..utils import construct_boto_client_retry_config


class S3Callback:
    def __init__(self, logger, bucket, key, filename, size):
        self._logger = logger
        self._bucket = bucket
        self._key = key
        self._filename = filename
        self._seen_so_far = 0
        self._size = size

    def __call__(self, bytes_amount):
        self._seen_so_far += bytes_amount
        percentage = (self._seen_so_far / self._size) * 100
        self._logger(
            f"Download of {self._bucket}/{self._key} to {self._filename}: {percentage}% complete"
        )


def construct_s3_client(
    max_attempts: int,
    region_name: Optional[str] = None,
    endpoint_url: Optional[str] = None,
    use_unsigned_session: bool = False,
    profile_name: Optional[str] = None,
    use_ssl: bool = True,
    verify: Optional[Union[str, bool]] = None,
    aws_access_key_id: Optional[str] = None,
    aws_secret_access_key: Optional[str] = None,
    aws_session_token: Optional[str] = None,
):
    check.int_param(max_attempts, "max_attempts")
    check.opt_str_param(region_name, "region_name")
    check.opt_str_param(endpoint_url, "endpoint_url")
    check.bool_param(use_unsigned_session, "use_unsigned_session")
    check.opt_str_param(profile_name, "profile_name")
    check.bool_param(use_ssl, "use_ssl")
    check.opt_str_param(profile_name, "aws_access_key_id")
    check.opt_str_param(profile_name, "aws_secret_access_key")
    check.opt_str_param(profile_name, "aws_session_token")

    client_session = boto3.session.Session(profile_name=profile_name)
    s3_client = client_session.resource(
        "s3",
        region_name=region_name,
        use_ssl=use_ssl,
        verify=verify,
        endpoint_url=endpoint_url,
        aws_access_key_id=aws_access_key_id,
        aws_secret_access_key=aws_secret_access_key,
        aws_session_token=aws_session_token,
        config=construct_boto_client_retry_config(max_attempts),
    ).meta.client

    if use_unsigned_session:
        s3_client.meta.events.register("choose-signer.s3.*", disable_signing)

    return s3_client
