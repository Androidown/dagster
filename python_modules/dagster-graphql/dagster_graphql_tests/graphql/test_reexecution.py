from dagster._core.test_utils import wait_for_runs_to_finish
from dagster_graphql.client.query import (
    LAUNCH_PIPELINE_EXECUTION_MUTATION,
    LAUNCH_PIPELINE_REEXECUTION_MUTATION,
)
from dagster_graphql.test.utils import execute_dagster_graphql, infer_job_selector

from .graphql_context_test_suite import ExecutingGraphQLContextTestMatrix
from .repo import csv_hello_world_ops_config

RUN_QUERY = """
query RunQuery($runId: ID!) {
  pipelineRunOrError(runId: $runId) {
    __typename
    ... on Run {
        status
      }
    }
  }
"""


class TestReexecution(ExecutingGraphQLContextTestMatrix):
    def test_full_pipeline_reexecution_fs_storage(self, graphql_context, snapshot):
        selector = infer_job_selector(graphql_context, "csv_hello_world")
        result_one = execute_dagster_graphql(
            graphql_context,
            LAUNCH_PIPELINE_EXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "runConfigData": csv_hello_world_ops_config(),
                    "mode": "default",
                }
            },
        )

        assert result_one.data["launchPipelineExecution"]["__typename"] == "LaunchRunSuccess"

        run_id = result_one.data["launchPipelineExecution"]["run"]["runId"]

        # overwrite run_id for the snapshot
        result_one.data["launchPipelineExecution"]["run"]["runId"] = "<runId dummy value>"
        result_one.data["launchPipelineExecution"]["run"]["runConfigYaml"] = (
            "<runConfigYaml dummy value>"
        )

        snapshot.assert_match(result_one.data)

        # reexecution
        result_two = execute_dagster_graphql(
            graphql_context,
            LAUNCH_PIPELINE_REEXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "runConfigData": csv_hello_world_ops_config(),
                    "executionMetadata": {
                        "rootRunId": run_id,
                        "parentRunId": run_id,
                    },
                    "mode": "default",
                }
            },
        )

        query_result = result_two.data["launchPipelineReexecution"]
        assert query_result["__typename"] == "LaunchRunSuccess"
        assert query_result["run"]["rootRunId"] == run_id
        assert query_result["run"]["parentRunId"] == run_id

    def test_full_pipeline_reexecution_in_memory_storage(self, graphql_context, snapshot):
        selector = infer_job_selector(graphql_context, "csv_hello_world")
        result_one = execute_dagster_graphql(
            graphql_context,
            LAUNCH_PIPELINE_EXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "runConfigData": csv_hello_world_ops_config(),
                    "mode": "default",
                }
            },
        )

        assert result_one.data["launchPipelineExecution"]["__typename"] == "LaunchRunSuccess"

        run_id = result_one.data["launchPipelineExecution"]["run"]["runId"]

        # overwrite run_id for the snapshot
        result_one.data["launchPipelineExecution"]["run"]["runId"] = "<runId dummy value>"
        result_one.data["launchPipelineExecution"]["run"]["runConfigYaml"] = (
            "<runConfigYaml dummy value>"
        )

        snapshot.assert_match(result_one.data)

        # reexecution
        result_two = execute_dagster_graphql(
            graphql_context,
            LAUNCH_PIPELINE_REEXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "runConfigData": csv_hello_world_ops_config(),
                    "executionMetadata": {
                        "rootRunId": run_id,
                        "parentRunId": run_id,
                    },
                    "mode": "default",
                }
            },
        )

        query_result = result_two.data["launchPipelineReexecution"]
        assert query_result["__typename"] == "LaunchRunSuccess"
        assert query_result["run"]["rootRunId"] == run_id
        assert query_result["run"]["parentRunId"] == run_id

    def test_pipeline_reexecution_successful_launch(self, graphql_context):
        selector = infer_job_selector(graphql_context, "no_config_job")
        result = execute_dagster_graphql(
            context=graphql_context,
            query=LAUNCH_PIPELINE_EXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "mode": "default",
                }
            },
        )

        assert result.data["launchPipelineExecution"]["__typename"] == "LaunchRunSuccess"
        assert result.data["launchPipelineExecution"]["run"]["status"] == "STARTING"

        wait_for_runs_to_finish(graphql_context.instance)

        run_id = result.data["launchPipelineExecution"]["run"]["runId"]
        result = execute_dagster_graphql(
            context=graphql_context, query=RUN_QUERY, variables={"runId": run_id}
        )
        assert result.data["pipelineRunOrError"]["__typename"] == "Run"
        assert result.data["pipelineRunOrError"]["status"] == "SUCCESS"

        # reexecution
        result = execute_dagster_graphql(
            context=graphql_context,
            query=LAUNCH_PIPELINE_REEXECUTION_MUTATION,
            variables={
                "executionParams": {
                    "selector": selector,
                    "executionMetadata": {
                        "rootRunId": run_id,
                        "parentRunId": run_id,
                    },
                    "mode": "default",
                }
            },
        )
        assert result.data["launchPipelineReexecution"]["__typename"] == "LaunchRunSuccess"

        wait_for_runs_to_finish(graphql_context.instance)

        new_run_id = result.data["launchPipelineReexecution"]["run"]["runId"]
        result = execute_dagster_graphql(
            context=graphql_context, query=RUN_QUERY, variables={"runId": new_run_id}
        )
        assert result.data["pipelineRunOrError"]["__typename"] == "Run"
        assert result.data["pipelineRunOrError"]["status"] == "SUCCESS"
