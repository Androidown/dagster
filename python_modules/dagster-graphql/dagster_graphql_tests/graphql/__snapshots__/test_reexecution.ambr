# serializer version: 1
# name: TestReexecution.test_full_pipeline_reexecution_fs_storage[0]
  dict({
    'launchPipelineExecution': dict({
      '__typename': 'LaunchRunSuccess',
      'run': dict({
        'mode': 'default',
        'pipeline': dict({
          'name': 'csv_hello_world',
        }),
        'resolvedOpSelection': None,
        'runConfigYaml': '<runConfigYaml dummy value>',
        'runId': '<runId dummy value>',
        'status': 'STARTING',
        'tags': list([
        ]),
      }),
    }),
  })
# ---
# name: TestReexecution.test_full_pipeline_reexecution_in_memory_storage[0]
  dict({
    'launchPipelineExecution': dict({
      '__typename': 'LaunchRunSuccess',
      'run': dict({
        'mode': 'default',
        'pipeline': dict({
          'name': 'csv_hello_world',
        }),
        'resolvedOpSelection': None,
        'runConfigYaml': '<runConfigYaml dummy value>',
        'runId': '<runId dummy value>',
        'status': 'STARTING',
        'tags': list([
        ]),
      }),
    }),
  })
# ---
